migration_scripts/bulk_load/src/footprint.clj

```diff
@@ -692,20 +802,86 @@
-        (into {} (for [[table stats] tables
-                       :let [extract (extract-kvs table stats)]
-                       :when (seq extract)]
-                   [table extract]))]
+        extracted (filter-map-table-stats extract-kv tables)]
 
     ;; Materialize `extracted` into a dictionary first, so `kv-size`
     ;; holds the estimated key-value store size (`for` produces a lazy
     ;; sequence).
     (assoc extracted :kv-store [{:self @kv-size}])))
+
+;; Abstract IDs and Addresses ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn abstracted-ids
+  "Simulate replacing 32B Addresses and Object IDs with an 8B key.
+
+  Requires creating a mapping from long IDs to short IDs"
+  [tables]
+  (let [object-ids
+        (get-in tables [:objects-snapshot 0 :tuples])
+
+        txr-distinct
+        (get-in tables [:tx-recipients 0 :cols :recipient :distinct])
+
+        txr-tuples
+        (get-in tables [:tx-recipients 0 :tuples])
+
+        ;; If the `distinct` stat is negative, its absolute value is
+        ;; interpreted as the denominator of a ratio.
+        addresses
+        (cond (not txr-distinct)        nil
+              (not (neg? txr-distinct)) txr-distinct
+              :else (* txr-tuples (/ (- txr-distinct))))
+
+        unique-ids
+        (+ (or object-ids 0) (or addresses 0))
+
+        ;; Pay 40B for each unique ID (32B for the long ID and 8B for
+        ;; the short ID), potentially twice if we don't have clustered
+        ;; tables.
+        abstract-ids
+        [{:self (* 40 unique-ids) :pkey (* 40 unique-ids)}]
+
+        replace-id-cols
+        (fn [table {:keys [tuples cols] :as stat}]
+          ;; If there is no column or tuple information, return the
+          ;; statistics unprocessed.
+          (if-not (and tuples cols)
+            stat
+            (loop [cols (keys cols) stat stat]
+              (if-let [col (first cols)]
+                (let [used (id-columns (fq-name table col))
+
+                      replace
+                      (fn replace
+                        ([sz-before] (replace sz-before 1))
+                        ([sz-before times]
+                         (let [debit (* 24 tuples times)]
+                           (cond
+                             (nil? sz-before) nil
+                             (> debit sz-before) 0
+                             :else (- sz-before debit)))))]
+                  (recur (rest cols)
+                         (cond
+                           ;; This column isn't an ID column, do nothing.
+                           (nil? used) stat
+
+                           ;; This is an ID column and one of its uses
+                           ;; is in the table's primary key.
+                           (neg? used)
+                           (-> stat
+                               (update :self replace)
+                               (update :pkey replace)
+                               (update :idx  replace (- (- used) 2)))
```

Comment:
why the `-2`?

Reply:
1 for the primary key, and 1 for the table itself.